h2/src/main/org/h2/table/TableView.java

```diff
@@ -80,41 +83,49 @@ public void replace(String querySQL,  Column[] newColumnTemplates, Session sessi
             boolean recursive, boolean force, boolean literalsChecked) {
         String oldQuerySQL = this.querySQL;
         Column[] oldColumnTemplates = this.columnTemplates;
-        boolean oldRecursive = this.recursive;
+        boolean oldRecursive = this.allowRecursive;
         init(querySQL, null,
                 newColumnTemplates == null ? this.columnTemplates
                         : newColumnTemplates,
-                session, recursive, literalsChecked);
+                session, recursive, literalsChecked, isTableExpression, isPersistent);
         DbException e = recompile(session, force, true);
         if (e != null) {
-            init(oldQuerySQL, null, oldColumnTemplates, session, oldRecursive, literalsChecked);
+            init(oldQuerySQL, null, oldColumnTemplates, session, oldRecursive, 
+                    literalsChecked, isTableExpression, isPersistent);
             recompile(session, true, false);
             throw e;
         }
     }
 
     private synchronized void init(String querySQL, ArrayList<Parameter> params,
-            Column[] columnTemplates, Session session, boolean recursive, boolean literalsChecked) {
+            Column[] columnTemplates, Session session, boolean allowRecursive, boolean literalsChecked, 
+            boolean isTableExpression, boolean isPersistent) {
         this.querySQL = querySQL;
         this.columnTemplates = columnTemplates;
-        this.recursive = recursive;
+        this.allowRecursive = allowRecursive;
         this.isRecursiveQueryDetected = false;
-        index = new ViewIndex(this, querySQL, params, recursive);
+        this.isTableExpression = isTableExpression;
+        this.isPersistent = isPersistent;
+        index = new ViewIndex(this, querySQL, params, allowRecursive);
         initColumnsAndTables(session, literalsChecked);
     }
 
-    private static Query compileViewQuery(Session session, String sql, boolean literalsChecked) {
+    private Query compileViewQuery(Session session, String sql, boolean literalsChecked, String viewName) {
         Prepared p;
-        session.setParsingView(true);
+        session.setParsingCreateView(true,viewName);
         try {
             p = session.prepare(sql, false, literalsChecked);
         } finally {
-            session.setParsingView(false);
+            session.setParsingCreateView(false,viewName);
         }
         if (!(p instanceof Query)) {
             throw DbException.getSyntaxError(sql, 0);
         }
-        return (Query) p;
+        Query q = (Query) p;
+        if(isTableExpression){
```

Comment:
Looks like here we have to `setNeverLazy(true)` only for recursive CTE, no?

Reply:
That works. Comitted.